### PR TITLE
Da: Network sampling

### PR DESCRIPTION
**Pull request by danielSanchezQ**

### CODE REVIEW

**Comment by danielSanchezQ** [comment]
Need of #688 to be merged first.

**Comment by danielSanchezQ** [comment]
@romanzac More stuff on DA networking (libp2p) here.

**Comment by romanzac** [comment]
> @romanzac More stuff on DA networking (libp2p) here.

I'll have more time to go through the code next week. libp2p is used for all types of messaging - consensus, da,  ?

**Comment by danielSanchezQ** [comment]
> > @romanzac More stuff on DA networking (libp2p) here.
> 
> I'll have more time to go through the code next week. libp2p is used for all types of messaging - consensus, da, ?

Yes, everything uses libp2p, although DA is probably the one that goes deeper in the stack.

**Comment by romanzac** [comment]
Hi, @danielSanchezQ I have a question about maximum subnet size. We were talking about 4096. I haven't seen any check on number of members in the implementations of trait MembershipHandler. It is kind of a restriction, so perhaps it will be added later ?

**Comment by danielSanchezQ** [comment]
> Hi, @danielSanchezQ I have a question about maximum subnet size. We were talking about 4096. I haven't seen any check on number of members in the implementations of trait MembershipHandler. It is kind of a restriction, so perhaps it will be added later ?

4096 is a theoretical limit (because of current hardware encoding times). We probably gonna go with just 2048 subnets for now. 
This is gonna be a configuration that should be in one of the upper layers (tbd). Or even in the final implementation of the MembershipHandler. I can ping you to the the code once its done :smiley:

File: nomos-da/network/core/src/sampling/behaviour.rs [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+        event: THandlerOutEvent<Self>,
+    ) {
+        let Either::Left(event) = event else {
+            unreachable!()
+        };
+        self.stream_behaviour
+            .on_connection_handler_event(peer_id, connection_id, event)
+    }
+
+    fn poll(
```
Comment (bacv) [inline nomos-da/network/core/src/sampling/behaviour.rs:491]: Would it be possible to reuse common parts from nomos behaviors to simplify `poll` and other network behavior methods?

**Review by bacv — APPROVED** [review]
Looks good!

File: nomos-da/network/core/src/sampling/behaviour.rs (hunk 1 above)
Comment (danielSanchezQ) [reply nomos-da/network/core/src/sampling/behaviour.rs:491]: I thought about it. Some patterns are really similar, and probably we could make some common functions. But the small differences (types, reactions, etc) may be a burden later on to maintain IMO.

**Pull request merged**